Registry with UTXO instead of PublicKey

linked0 opened the pull request:
When referring to the nodes individually, we should then use the UTXO instead of the public key because public keys are not currently unique.

Relates to #1309

ferencdg commented on source/agora/registry/API.d:
```diff
@@ -43,6 +47,25 @@ public struct RegistryPayload
+
+    ***************************************************************************/
+
+    static public Hash makeRegistryKey (in PublicKey pubkey, in UTXO utxo)
```
do we really need to has the pubkey and utxo together? utxo has an output, and if the utxo is a payment transaction (as opposed to be a script), then the output should contain the pubkey(or its hash)

linked0 replied:
Good point! But there are some cases where `outsider_validators` don't yet have UTXOs in unit tests. I will consider another way to deal with it. Ty.

ferencdg commented on source/agora/node/Validator.d:
```diff
@@ -449,6 +449,21 @@ public class Validator : FullNode, API
+        if (hash != Hash.init)
+        {
+            UTXO utxo;
+            this.utxo_set.peekUTXO(hash, utxo);
```
peekUTXO has a bool return value to indicate if the peeking was successful

linked0 replied:
What do you mean? I used the function to get a UTXO for the hash.

ferencdg replied:
> What do you mean? I used the function to get a UTXO for the hash.

what if 
this.utxo_set.peekUTXO(hash, utxo); 
returns false? it would mean 
utxo == UTXO.init and we would return that

linked0 replied:
You mean the line of `if (hash != Hash.init)` is not needed?

linked0 replied:
I will check the returned value with assert. Ty.

ferencdg commented on source/agora/node/Validator.d:
```diff
@@ -449,6 +449,21 @@ public class Validator : FullNode, API
+    public override UTXO getValidatorUTXO (in PublicKey pubkey) @trusted nothrow
+    {
+        auto hash = this.enroll_man.validator_set.getEnrolledUTXO(
+                this.ledger.getBlockHeight(), pubkey);
```
here we are referring to the block height as 'this.ledger.getBlockHeight()' - but how would this work with a validator that is catching up, and still doesn't have the correct hight?

linked0 replied:
I don't think it could be a big problem because even if there are multiple `RegistryPayload`s for a `PublicKey` in the registry, they will indicate the same address. I hope this explanation makes sense to you. :)

ferencdg replied:
> I don't think it could be a big problem because even if there are multiple `RegistryPayload`s for a `PublicKey` in the registry, they will indicate the same address. I hope this explanation makes sense to you. :)

they can actually indicate different addresses, if the validator's IP changed :) However we do authenticate validators during the handshake, so I agree this is not a big problem.

ferencdg commented on source/agora/node/FullNode.d:
```diff
@@ -534,6 +535,24 @@ public class FullNode : API
+
+    public UTXO getValidatorUTXO (in PublicKey pubkey) @trusted nothrow
+    {
+        assert(0);
```
just a small thing: can we have assert(0, "some string here") - when we will switch to release build we might not have access to the stack trace, so we might not know what the actual error was

linked0 replied:
Updated. There was a mistake. The code in `Validator` was in fact moved into this function because the `Validator` doesn't need to have it. Ty.

ferencdg commented on source/agora/registry/Server.d:
```diff
@@ -84,19 +83,22 @@ public final class NameRegistry: NameRegistryAPI
 
     public override void putValidator (RegistryPayload registry_payload)
```
I was thinking a bit more about this:
let's say an attacker(node A) with a pubkey A signs the payload and the payload contains
1. the attacker pubkey A
2. UTXO X where UTXO X is actually a UTXO for node B

Currently we only check if the payload is properly signed, but we never check if UTXO X belongs to node B or node A

For this reason it might make sense to include the pubkey into that hash(or add the extra check as above which is what I think we should do)

Geod24 commented on source/agora/registry/API.d:
```diff
@@ -43,6 +47,25 @@ public struct RegistryPayload
     /// signature over the `data` member
     public Signature signature;
 
+    /***************************************************************************
+
+        make a key for a `RegistryPayload`
+
+        Params:
+            pubkey = the address used to make a key for a `RegistryPayload`
+            utxo = the UTXO used to make a key for a `RegistryPayload`
+
+        Returns:
+            a key of `Hash` type
+
+    ***************************************************************************/
+
+    static public Hash makeRegistryKey (in PublicKey pubkey, in UTXO utxo)
+        @safe nothrow @nogc
+    {
+        return hashMulti(pubkey, utxo);
+    }
+
```
I don't understand why we hash both together. The UTXO hash should be enough.

linked0 replied:
It's because there are cases where `outsider` validators don't have any UTXO in unit tests.

ferencdg replied:
outsider validators shouldn't register themselves I think

linked0 replied:
Currently, the outsider validators register themselves. But I know we have a plan to remove outsider validators in our test codes.

Geod24 commented on source/agora/registry/API.d:
```diff
@@ -24,7 +25,10 @@ import vibe.web.rest;
 public struct RegistryPayloadData
 {
-    /// the public key that we want to register
+    /// the UTXO that we want to register
+    public UTXO utxo;
+
+    /// the public key that we want to register if there's no UTXO
     public PublicKey public_key;
```
Since the `UTXO` contains the `PublicKey`, the second field is redundant.

ferencdg replied:
I also mentioned this in a previous comment, but if we use only the UTXO as the key, we might need to do some extra checks: https://github.com/bosagora/agora/pull/2080#discussion_r631492163

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -325,6 +325,43 @@ public class ValidatorSet
+                "PublicKey for enrollment: {}", ex.msg, pubkey);
+            assert(0);
+        }
+    }
```
We shouldn't need such a function. Its presence indicates that the switch is not complete, and we're still using `PublicKey` instead of `UTXO` / `Hash` in some places.
Additionally, it turns a "recoverable" error (`Exception`) into an unrecoverable one (`assert(0)`). I think we should leave the caller of this function decide whether or not the error is recoverable.

linked0 replied:
I will implement another approach soon without introducing a new function like this.

Geod24 commented on source/agora/flash/Node.d:
```diff
@@ -1595,10 +1595,17 @@ public class AgoraFlashNode : FlashNode
+            if(utxo == UTXO.init)
+            {
+                log.trace("Couldn't find the UTXO for this node: {}.", pair.key);
+            }
```
That looks buggy. If we can't find the `utxo`, we send a payload with `UTXO.init`.

Geod24 reviewed (commented):
This does 80% of the change, I believe. I'll have a look at it locally.

linked0 commented:
How about this approach? Please leave your opinions. I will fix some nits anyway.

@Geod24, I think we need to have a function to get UTXOs for a public key because the `required_peer_keys` parameter of the `discover` function of `NetworkManager` is a set of `PublicKey`s. Do you think we should change the `required_peer_keys` should be `Set!Hash(UTXO)` instead of `Set!PublicKey`?

Geod24 commented on source/agora/flash/Node.d:
```diff
@@ -1595,10 +1595,17 @@ public class AgoraFlashNode : FlashNode
 
         foreach (pair; this.getManagedKeys())
         {
+            UTXO utxo = this.network.getEnrollmentUTXO();
+            if(utxo == UTXO.init)
+            {
+                log.trace("Couldn't find the UTXO for this node: {}.", pair.key);
+            }
+
```
Urgh, the flash layer really shouldn't piggyback on the registry here... I think this will break the flash layer big time.

linked0 replied:
Ah, Sorry. This is the intermediate code, not the final code. I just wanted to upload my intermediate code into GitHub.

linked0 commented:
I'm working on the issue by starting to add the following code in the third commit.
```
package(scpd)
struct UTXOHash {
    extern(D) this(uint512 value) @safe pure nothrow @nogc
    {
        this.utxo_hash_ = value;
    }

    extern(D) this(typeof(this.tupleof) args) @safe pure nothrow @nogc
    {
        this.tupleof = args;
    }

    int32_t type_;
    uint512 utxo_hash_;
    alias utxo_hash_ this;
}

... // Omitted
alias NodeID = UTXOHash;
```
By adding the above code and the related code, there are so many errors. And I'm working on the errors.

@Geod24  Could you leave your opinion about the third commit in this PR and give me some guidance about how to change code in the Stellar part?
I'm wondering I'm going in the right direction on this issue. It would be good to have a whiteboard soon.

Geod24 commented on source/agora/common/Config.d:
```diff
@@ -939,15 +939,15 @@ private auto get (string section, string name, Converter) (
 {
     import std.conv;
-    import scpd.types.Stellar_types : uint256, NodeID;
+    import scpd.types.Stellar_types : uint256, uint512, NodeID;
```
You can remove the uint256 import I think.

Geod24 commented on source/agora/node/Validator.d:
```diff
@@ -143,8 +143,8 @@ public class Validator : FullNode, API
-        buildRequiredKeys(this.config.validator.key_pair.address, this.qc,
-            this.required_peer_keys);
+        buildRequiredKeys(this.config.validator.key_pair.address,
+            this.qc, this.required_peer_keys, this.required_peer_utxos);
```
I think we can / should remove `required_peer_keys`

Geod24 commented on source/scpd/scp/SCPDriver.d:
```diff
@@ -148,11 +148,11 @@ nothrow:
         // `toStrKey` returns StrKey encoded string representation
         //std::string toStrKey(ref const(PublicKey) pk, bool fullKey = true) const;
-        void* toStrKey(ref const(PublicKey) pk, bool fullKey = true) const;
+        void* toStrKey(ref const(UTXOHash) pk, bool fullKey = true) const;
```
Don't think that's going to work, this is for keys, not hashes

Geod24 commented on source/scpd/types/Stellar_types.d:
```diff
@@ -68,6 +72,23 @@ struct PublicKey {
     alias ed25519_ this;
 }
 
+package(scpd)
+struct UTXOHash {
+    extern(D) this(uint512 value) @safe pure nothrow @nogc
+    {
+        this.utxo_hash_ = value;
+    }
+
+    extern(D) this(typeof(this.tupleof) args) @safe pure nothrow @nogc
+    {
+        this.tupleof = args;
+    }
+
+    int32_t type_;
+    uint512 utxo_hash_;
+    alias utxo_hash_ this;
+}
```
I think this should be handled one level higher, e.g. the `PublicKey` type has to go.

Geod24 reviewed (commented):
Lots of good stuff, but I think we might need to abstract things away on the SCP side a bit more. Stellar used to have all the crypto operations on the SCP layer, they abstracted it away somewhere between v11.2.0 and v16.0.0, but there are some lingering types.

linked0 commented:
Ready for review.
Issue #1309  needs to make the flash nodes use UTXO also instead of PublicKey as the next phase. I will be working on that and submit another issue.

Geod24 commented:
So my recent foray into SCP came from this, as I realized we could greatly simplify our interfacing.
One other thing that came up while reviewing this is that the Flash layer piggyback on the key registration of the registry, which is problematic for us, because it means 1) this change cannot be completed, and 2) the registry is an obvious DoS target.

Geod24 commented on source/agora/consensus/state/ValidatorSet.d:
```diff
@@ -55,7 +55,7 @@ public alias EnrollmentFinder = bool delegate (in Hash enroll_key, out Enrollmen
 /// A Height and PublicKey pair to represent expiring Validators
 public alias ExpiringValidator = Tuple!(Height, "enrolled_height",
-    PublicKey, "pubkey");
+    Hash, "utxo", PublicKey, "pubkey");
```
Can you extract those changes (add8fd7a6a59f589abb60561dece48298e3e98bd / source/agora/consensus/state/ValidatorSet.d) in their own PR ? We can merge it immediately.

linked0 replied:
Sure, I will.

Geod24 commented:
Pretty nice BTW! The first commit does point to a few issues we have, but with a few SCP changes, it should look much nicer.

Geod24 commented:
Can you submit the first commit as its own PR ? Then we can just switch SCP.

linked0 commented:
> Can you submit the first commit as its own PR ? Then we can just switch SCP.

Sure.

linked0 committed:
Registry with UTXO instead of PublicKey

We use UTXOs to register network addresses instead of PublicKey,which makes users run multiple nodes with the same public key.

linked0 committed:
Remove the unnecessary `key` from NodeConnInfo

Geod24 commented:
After more research and discussion, we decided not to go this way.
